src/containers/memory_arena.c

```diff
@@ -1,67 +1,138 @@
+    arena->current_block = arena->first_block;
+    arena->current_adr = arena->current_block->memory_start;
+    arena->n_blocks = 1;
+    
     return arena;
+    
+    error:
+    if (arena != NULL) {
+        if (arena->first_block != NULL) memory_free(arena->first_block);
+        memory_free(arena);
+    }
+    
+    return NULL;
 }
 
 
 internal void
 memory_arena_destroy(MemoryArena* arena) {
     check(arena != NULL, "arena is NULL");
     
-    arena->size = 0;
-    arena->current = 0;
-    arena->memory = NULL;
+    MemoryArenaBlock* it = arena->first_block;
+    MemoryArenaBlock* aux = NULL;
+    
+    while (it != NULL) {
+        aux = it;
+        it = it->next;
+        memset(aux, 0, sizeof(*aux) + arena->block_size);
+        memory_free(aux);
+    }
+    memset(arena, 0, sizeof(*arena));
     memory_free(arena);
     
     error:
     return;
 }
 
 
+internal void
+memory_arena_clear(MemoryArena* arena) {
+    check(arena != NULL, "arena is NULL");
+    
+    arena->current_block = arena->first_block;
+    arena->current_adr = arena->first_block->memory_start;
+    
+    error:
+    return;
+}
+
+
+
 internal void*
 memory_arena_push(MemoryArena* arena, sz size) {
     u8* memory = NULL;
+    MemoryArenaBlock* block = NULL;
+    
     check(arena != NULL, "arena is NULL");
     check(size != 0, "size is 0");
+    check(size <= arena->block_size,
+          "size %zu is bigger than the block size %zu",
+          size, arena->block_size);
+    // NOTE: We could like allocate into multiple block but discontinuities in memory are 
+    // hard to work with, for now just increase the block size when this happens
     
     // NOTE: On this machine if memory is not aligned based on the type, its okay, but
     // NOTE: pretty sure that the CPU is doing more that it should
     // NOTE: for this, aligned everything to sizeof(void*) so that there is no problem
+    
+    // TODO: we could make this better by considering types that are lower than the size_t
+    // TODO: lets see if it will make sense to do this
     sz reminder = size % sizeof(void*);
     if (reminder != 0) {
         size += sizeof(void*) - reminder;
     }
     
-    // TODO: should it be with =?
-    if (arena->current + size > arena->size) {
-        log_warning("The memory arena cannot hold another %zu bytes", size);
+    // NOTE: check if it fits in current block
+    if (arena->current_adr + size <= arena->current_block->memory_end) {
+        memory = arena->current_adr;
+        arena->current_adr += size;
     } else {
-        memory = arena->memory + arena->current;
-        arena->current += size;
+        // NOTE: check if the current block has a next and use that
+        if (arena->current_block->next != NULL) {
+            arena->current_block = arena->current_block->next;
+        } else {
+            // NOTE: need to alocate another block and move the adr to it
+            block = memory_arena_block_create(arena->block_size);
```

Comment:
Add a function for adding a new block in the arena because its the same code in the create

Reply:
Done